Remove basic feature checks from license state

rjernst opened the pull request:
The XPackLicenseState is a utility to handle checking the currently
configured license level against the required license level of each
licensed feature. Originally all licensed features were paid features
and licenses always had a limited time scope. However, since the basic
license was introduced, all users have an unlimited time use of basic
features; there is no actual "basic license", rather it is the base case
for the default distribution. Therefore, the checks on basic features in
license state are no-ops because the current license always at least
allows basic features, and the basic license cannot be expired.

This commit removes all of the features from the license state marked as
BASIC or MISSING (a level lower than BASIC that also predates the basic
license and is no longer relevant). It also adds an assertion that no
basic license features can be added to the license state in the future.

rjernst committed:
Remove basic feature checks from license state

The XPackLicenseState is a utility to handle checking the currently
configured license level against the required license level of each
licensed feature. Originally all licensed features were paid features
and licenses always had a limited time scope. However, since the basic
license was introduced, all users have an unlimited time use of basic
features; there is no actual "basic license", rather it is the base case
for the default distribution. Therefore, the checks on basic features in
license state are no-ops because the current license always at least
allows basic features, and the basic license cannot be expired.

This commit removes all of the features from the license state marked as
BASIC or MISSING (a level lower than BASIC that also predates the basic
license and is no longer relevant). It also adds an assertion that no
basic license features can be added to the license state in the future.

elasticmachine commented:
Pinging @elastic/es-security (:Security/License)

rjernst committed:
remove forgotten analytics constant

rjernst committed:
Merge branch 'master' into refactor_license15

rjernst committed:
checkstyle

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/filter/SecurityActionFilter.java:
```diff
@@ -69,17 +69,6 @@ public SecurityActionFilter(AuthenticationService authcService, AuthorizationSer
-         A functional requirement - when the license of security is disabled (invalid/expires), security will continue
-         to operate normally, except all read operations will be blocked.
-         */
-        if (licenseState.checkFeature(Feature.SECURITY_STATS_AND_HEALTH) == false && LICENSE_EXPIRATION_ACTION_MATCHER.test(action)) {
-            logger.error("blocking [{}] operation due to expired license. Cluster health, cluster stats and indices stats \n" +
-                    "operations are blocked on license expiration. All data operations (read and write) continue to work. \n" +
-                    "If you have a new license, please update it. Otherwise, please reach out to your support contact.", action);
-            throw LicenseUtils.newComplianceException(XPackField.SECURITY);
-        }
```
My understanding on the intention of this piece of code is that we do not encourage users to run with expired commerical licenses, e.g. PLATINUM, forever. An expired PLATINUM license is still allowed to access some platinum features, e.g. DLS/FLS, for security reasons. If an expired license can access all BASIC featues, plus a fair bit more, this would put our licensing in an awkward situation. Therefore, we disable stats and health APIs to prompt users to either downgrade to a BASIC or renew the subscription.

In fact, many of the BASIC features have similar behaviour as the stats and health API, e.g. `ROLLUP(OperationMode.MISSING, true)`. The feature is available in BASIC, but also requires the license to be active.

ywangd reviewed (commented):
I have one main concern:

Currently many of the basic features require the license to be active. My understanding is that we do not encourage users to run with expired commerical licenses forever. An expired PLATINUM license is still allowed to access some platinum features, e.g. DLS/FLS, for security reasons. If an expired license can access all BASIC featues, plus a fair bit more, this would put our licensing in an awkward situation. Therefore, we disable some of the BASIC features so that user can choose to either downgrade to a BASIC license or renew the subscription. With changes in this PR, an expired PLATINUM license would be all around better than a BASIC. Is this our intention?

I also have a few other less important suggestions:

* With this change, there are many `licenseState` fields that can be removed from various TransportAction classes, e.g. `RollupInfoTransportAction` as well as `ApiKeyService`.
* There seems to be opportunities for extracting common behaviours. For example, the `XxxInfoTransportAction` classes, since BASIC features can no longer be dsiabled and is always allowed, maybe we could have a superclass such as `BasicFeatureInfoTransportAction`. Security is a notable anomaly here. I wish it can also be always enabled, but that's a separate story.
* For the `Usage` classes, e.g. `RollupFeatureSetUsage`, I wonder whether we could add another constructor:
```java
public RollupFeatureSetUsage() {
    super(XPackField.ROLLUP, true, true);
}
```
and mark the current constructor `public RollupFeatureSetUsage(boolean available)` as deprecated or  just remove it. So we are not relying on the callers to pass in the right value.

rjernst committed:
Add back disabling of health and stats on expired license

rjernst committed:
checkstyle

rjernst committed:
Merge branch 'master' into refactor_license15

rjernst committed:
more merge cleanup

rjernst committed:
address some feedback

rjernst committed:
remove transform aux function

rjernst committed:
cleanup unused licenseState

rjernst committed:
spotless

rjernst committed:
checkstyle

rjernst committed:
test tweak

rjernst committed:
Merge branch 'master' into refactor_license15

rjernst committed:
fix tests

rjernst committed:
remove mutations from spatial feature set usage

rjernst commented:
@elasticmachine run elasticsearch-ci/2

rjernst committed:
Merge branch 'master' into refactor_license15

rjernst committed:
fix ML transforms with remote clusters

rjernst commented:
@ywangd I've brought this PR back up to date and addressed your comments. Can you take another look?

Re your broader comment:
> With changes in this PR, an expired PLATINUM license would be all around better than a BASIC. Is this our intention?

Yes, this is the intention.

rjernst committed:
Merge branch 'master' into refactor_license15

ywangd commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/rest/action/user/RestHasPrivilegesActionTests.java:
```diff
@@ -38,6 +40,7 @@
      */
     public void testBodyConsumed() throws Exception {
         final XPackLicenseState licenseState = mock(XPackLicenseState.class);
+        when(licenseState.isSecurityEnabled()).thenReturn(true);
```
To retain the existing behaviour of this test, should this mock return `false` instead of `true` so that `action.handleRequest(...)` does not throw exception but returns a response about security is not enabled (currently it returns security not available due to license)?

ywangd replied:
We discussed this in another channel. After another look, I think the change is in fact closer to what this test tries to assert:
1. Send a request without an username
2. It should _not_ fail with IllegalArgumentException complaining about unconsumed payload

Both the existing version and the change do the job because both of them will run through `BaseRestHandler#handleRequest`, where the unconsumed body is checked.
The existing version does its job by ensure there is no exception. This relies on the fact that “security disabled” message takes precedence over the “no authenticated user” exception. The new change enables security so that the “no authenticated user” exception now bubbles up. Both “security disabled” message and “no authenticated user” exception have lower priority than the “unconsumed body” exception. So the test does its job if either of them can be asserted. But the new change has the advantage of an additional assertion on the exception to be “no authenticated user”, which is more to the point.

In summary, the new change does its job slightly differently than the existing version. But I think it's a small improvement. So we can discard my previous comment.

ywangd commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/SqlUsageTransportAction.java:
```diff
@@ -58,7 +57,7 @@ protected void masterOperation(Task task, XPackUsageRequest request, ClusterStat
                 .collect(Collectors.toList());
             Counters mergedCounters = Counters.merge(countersPerNode);
-            SqlFeatureSetUsage usage = new SqlFeatureSetUsage(available, mergedCounters.toNestedMap());
+            SqlFeatureSetUsage usage = new SqlFeatureSetUsage(true, mergedCounters.toNestedMap());
```
Yet another `available` parameter.

ywangd commented on x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/TransformUsageTransportAction.java:
```diff
@@ -96,7 +90,7 @@ protected void masterOperation(
 
         ActionListener<TransformIndexerStats> totalStatsListener = ActionListener.wrap(statSummations -> {
-            var usage = new TransformFeatureSetUsage(available, transformsCountByState, statSummations);
+            var usage = new TransformFeatureSetUsage(true, transformsCountByState, statSummations);
```
Here again, the `available` parameter.

ywangd commented on x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/VectorsUsageTransportAction.java:
```diff
@@ -76,7 +70,7 @@ protected void masterOperation(Task task, XPackUsageRequest request, ClusterStat
         }
         VectorsFeatureSetUsage usage =
-            new VectorsFeatureSetUsage(vectorsAvailable, numDenseVectorFields, avgDenseVectorDims);
+            new VectorsFeatureSetUsage(true, numDenseVectorFields, avgDenseVectorDims);
```
The `available` parameter again.

ywangd commented on x-pack/plugin/voting-only-node/src/main/java/org/elasticsearch/cluster/coordination/VotingOnlyNodeFeatureSet.java:
```diff
@@ -53,36 +43,26 @@ public boolean enabled() {
-            final boolean available = licenseState.checkFeature(Feature.VOTING_ONLY);
-            final VotingOnlyNodeFeatureSetUsage usage =
-                new VotingOnlyNodeFeatureSetUsage(available);
+            final VotingOnlyNodeFeatureSetUsage usage = new VotingOnlyNodeFeatureSetUsage(true);
```
I believe this is the last occurrence of the `available` parameter.

ywangd reviewed (approved):
LGTM. I left some comments, nothing really major.

> > With changes in this PR, an expired PLATINUM license would be all around better than a BASIC. Is this our intention?
>>
> Yes, this is the intention.
>
Well almost, the health and stats related actions are still blocked for expired licenses. So it still creates some prompt for users to choose either renew license or explicitly go back to basic.

I kinda like having a central list where every feature is explicitly listed along with its license requirement. It serves as a good reference. If we configure the license requirement correctly, something like `...(OperationMode.MISSING, false)`, we can even leave all the call sites of `checkFature` unchanged and just change the method itself to not track usages for any features of BASIC or lower. It also enforces license requirement be specified in one place so there is no place for inconsistency (among different classes usage/info/action etc). With that being said, I do like how much code we can delete because of the relevant Feature Enums are deleted. Given people who develop any new feature should also be responsible or consciously involved in license decision, I think it is reasonably to expect consistency will be taken care of. Overall I am fine with this approach.

rjernst committed:
Merge branch 'master' into refactor_license15

rjernst committed:
more review comments

rjernst committed:
checkstyle

rjernst committed:
fix compile
